### DISCUSSION TITLE
Git and Git Subrepo

### DISCUSSION THREAD

**Discussion by tanzeelraza** [community/community#22875] [New to GitHub]
<p>Hi,</p>
<p>I am newbie to Git and starting to learn it. I have been using SVN for quite a long time. I want to simulate following.</p>
<ol>
<li>I have a main repo in folder ‘main’. I can pull, commit and push in this folder.</li>
<li>I have one folder in ‘main’ named ‘srepo’. I am able to clone a subrepo in this folder ‘srepo’</li>
</ol>
<p>I added a few files to subrepo ‘srepo’ locally and want to push to upstream subrepo. when I execute command ‘git subrepo add FileName.txt’  then it gives me error git-subrepo: ‘add’ is not a command. See ‘git subrepo help’.</p>
<p>When I execute command ’ <strong>git subrepo commit</strong>’ it gives me following message:-</p>
<p>git-subrepo: Command ‘commit’ requires arg ‘subdir’.<br>
  Usage: git subrepo commit &lt;subdir&gt; [&lt;subrepo-ref&gt;]<br>
  Add subrepo branch to current history as a single commit.<br>
  This command is generally used after a hand-merge. You have done a <code>subrepo&lt;/font&gt;   &lt;font color="#0000FF"&gt;  branch</code> and merged (rebased) it with the upstream. This command takes the<br>
  HEAD of that branch, puts its content into the subrepo subdir and adds a new<br>
  commit for it to the top of your mainline history.</p>
<p>  This command requires that the upstream HEAD be in the <code>subrepo/&amp;lt;subdir&amp;gt;</code><br>
  branch history. That way the same branch can push upstream. Use the<br>
  <code>--force</code> option to commit anyway.</p>
<p>  The <code>commit</code> command accepts the <code>--force</code> option.</p>
<p>After adding subdir name with commit command it says “git-subrepo: Can’t commit: ‘subrepo/origin’ doesn’t contain upstream HEAD.”</p>
<p>Please let me know how fix this?</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>The main repo in folder <code>main</code> would be a standard repository. The folder <code>main/srepo</code> would be what Git calls a <a href="https://git-scm.com/docs/gitsubmodules" rel="nofollow noopener">submodule</a>. But because submodule working trees within a superproject are tied to a specific commit, you wouldn’t typically develop directly within <code>main/srepo</code> but you would have a <code>srepo</code> project somewhere else that you would develop in and then update the commit that <code>main</code> has to that known-good version.</p>
<p>If you don’t want to have that extra formalism of the submodule, then I would recommend just treating <code>main/srepo</code> as a component that you work on logically separately but isn’t phyisically separated. You could do this by having just one Git repository and working on separate branches that only touch things inside <code>main/srepo</code> <strong>or</strong> outside <code>main/srepo</code> but not both.</p>
<p>I hope that helps!</p>

**Comment by sesam** [comment]
<p>Hey, <a href="https://github.com/ingydotnet/git-subrepo" rel="nofollow noopener">https://github.com/ingydotnet/git-subrepo</a> is a thing, and it’s not (really) submodules. It’s also tricky to figure out how to use.</p>
<p>Starting out:<br>
git subrepo clone # clone existing repo. Use --force to overwrite into an already existing folder<br>
or:<br>
git subrepo init # make a repo out of a previously existing and possibly not empty folder. Remote repo url can be supplied later during push, saved with --save</p>
<p>Daily usage:</p>
<p>git subrepo pull # see <code>git subrepo help</code> for a list of all the steps this command does. It’s complicated.<br>
git subrepo push # push changes. But always pull first (says subrepo docs…)</p>
